### DISCUSSION TITLE
Is it possible function calls with streaming

### DISCUSSION THREAD

**Discussion by emrediga** [marcominerva/ChatGptNet#152] [Ideas]
Hi,

If the response message is too long, we need to wait for all the responses to be downloaded from the AskAsync method. But there is no delay in the AskStreamAsync method. 

Problem is if we want to use function calls we have to use AskAsync Method. 

Is it possible to implement function calls in AskStreamAsync method?

**Comment by marcominerva** [comment]
In function calling, we have a response in JSON format. So, we need to wait until the message is complete to be sure the response JSON is valid.

**Reply by emrediga** [reply]
Maybe you should let the developer validate the json. I can implement this approach by making a small change to your library and there will be no lag or latency in the UI. But the problem is that chatgpt sends the function name first. Then sends the arguments. 

This is the first response of the stream.

```
{
    "id": "chatcmpl-9413oekZI2p7TjDj50NiNcQmwDGqr",
    "object": "chat.completion.chunk",
    "created": 1710744416,
    "model": "gpt-4-0125-preview",
    "system_fingerprint": "fp_ace12ee232",
    "choices": [
        {
            "index": 0,
            "delta": {
                "role": "assistant",
                "content": null,
                "function_call": {
                    "name": "GetCurrentWeather",
                    "arguments": ""
                }
            },
            "logprobs": null,
            "finish_reason": null
        }
    ]
}
```

Other responses comes with only arguments.

```
{
    "id": "chatcmpl-9413oekZI2p7TjDj50NiNcQmwDGqr",
    "object": "chat.completion.chunk",
    "created": 1710744416,
    "model": "gpt-4-0125-preview",
    "system_fingerprint": "fp_ace12ee232",
    "choices": [
        {
            "index": 0,
            "delta": {
                "function_call": {
                    "arguments": "{\""
                }
            },
            "logprobs": null,
            "finish_reason": null
        }
    ]
}
```

**Reply by marcominerva** [reply]
If you want to make a PR to add this feature, you're more then welcome. For what concerns the problem about the function name that is sent first, you could follow an approach similar to the one I use to attach the `PromptFilterResult` to every response streaming chunk. Take a look at https://github.com/marcominerva/ChatGptNet/blob/master/src/ChatGptNet/ChatGptClient.cs#L128.

**Reply by emrediga** [reply]
I am not sure if i am doing right. But here is the changes that i made.  Result should be like in the gif.
I also had to modify some extension methods. Can u review the solution and give feedback if it is ok or not?

[ChatGptNet-develop.zip](https://github.com/marcominerva/ChatGptNet/files/14634698/ChatGptNet-develop.zip)

![ezgif-3-51f2847604 1](https://github.com/marcominerva/ChatGptNet/assets/73991881/4bc9a366-bdb9-4724-94a1-ce729190c0f0)

**Reply by marcominerva** [reply]
Can you make a PR with your changes?

**Reply by emrediga** [reply]
I have tried many times but company laptop does not allow me to connect github.

**Reply by marcominerva** [reply]
What do you mean? You cannot create a fork of the repository and push to it?

**Reply by emrediga** [reply]
Company strategy but i did it. I had to ignore firewall and antivirus :D
